src/shogun/machine/GLM.cpp

```diff
@@ -0,0 +1,145 @@
+#include <shogun/optimization/ConstLearningRate.h>
+
+#include <cmath>
+
+using namespace shogun;
+
+
+
+
+
+GLM::GLM()
+{
+	SG_ADD(&distribution, "distribution_type", "variable to store name of distribution type", ParameterProperties::HYPER);
+	SG_ADD(&m_eta, "eta", "threshold parameter that linearizes the exp() function above eta", ParameterProperties::HYPER);
+	SG_ADD(&m_lambda, "lambda", "regularization parameter of penalty term", ParameterProperties::HYPER);
+	SG_ADD(&m_alpha, "alpha", "weighting between L1 penalty and L2 penalty term of the loss function", ParameterProperties::HYPER);
+	SG_ADD(&m_tolerance, "tolerance", "convergence threshold or stopping criteria", ParameterProperties::HYPER);
+	
+	m_gradient_updater = std::make_shared<GradientDescendUpdater>();
+	m_learning_rate = std::make_shared<ConstLearningRate>();
+	m_penalty = std::make_shared<ElasticNetPenalty>();
+	m_cost_function = std::make_shared<GLMCostFunction>();
+}
+
+GLM::GLM(GLM_DISTRIBUTION distr, float64_t alpha, float64_t lambda, float64_t learning_rate, int32_t max_iterations, float64_t tolerance, float64_t eta): GLM()
+{
+	distribution=distr;
+	m_alpha=alpha;
+	m_lambda=lambda;
+	m_max_iterations=max_iterations;
+	m_tolerance=tolerance;
+	m_eta=eta;
+
+	m_learning_rate->set_const_learning_rate(learning_rate);
+
+	m_penalty->set_l1_ratio(m_alpha);
+}
+
+std::shared_ptr<RegressionLabels> GLM::apply_regression(std::shared_ptr<Features> data)
+{
+	LinearMachine::set_features(std::static_pointer_cast<DotFeatures>(data));
+
+	if (!LinearMachine::features)
+		return std::make_shared<RegressionLabels>(SGVector<float64_t>());
+
+	auto num = LinearMachine::features->get_num_vectors();
+	ASSERT(num>0)
+	ASSERT(m_w.vlen==features->get_dim_feature_space())
+	SGVector<float64_t> out(num);
+	LinearMachine::features->dense_dot_range(out.vector, 0, num, NULL, m_w.vector, m_w.vlen, bias);
+	auto result = m_cost_function->non_linearity(out);
+	
+	return std::make_shared<RegressionLabels>(result);
+}
+
+void GLM::init_model(const std::shared_ptr<Features>& data)
+{
+	ASSERT(m_labels)
+	if (data)
+	{
+		if (!data->has_property(FP_DOT))
+			error("Specified features are not of type CDotFeatures");
+		LinearMachine::set_features(std::static_pointer_cast<DotFeatures>(data));
+	}
+	ASSERT(features)
+
+	NormalDistribution<float64_t> normal_dist;
+	auto n_features = LinearMachine::features->get_dim_feature_space();
+
+	if(m_w.vlen == 0)
+	{
+		if (m_compute_bias && bias == 0)
+			bias = 1.0 / (n_features + 1) * normal_dist(m_prng);
+
+		if(n_features > 0)
+		{
+			m_w = SGVector<float64_t>(n_features);
+			for (int i = 0; i < n_features; i++)
+			{
+				auto rand = normal_dist(m_prng);
+				m_w[i] = 1.0 / (n_features + 1) * rand;
+			}
+		}
+	}
+	m_cost_function->set_target(shared_from_this()->as<GLM>());
+
+}
+
+void GLM::iteration()
+{
+	auto learning_rate = m_learning_rate->get_learning_rate(m_current_iteration);
+	SGVector<float64_t> w_old(m_w.vlen);
+	for(int i = 0; i<m_w.vlen; i++)
```

Comment:
you can copy the whole memory block at once using e.g. sg_memcpy

Reply:
I would say use `clone` no? It will handle allocation + copying

Reply:
ah yes `SGVector::clone`